dashboard/app/config_linux_test.go

```diff
@@ -0,0 +1,186 @@
+mkdirat(r0, &(0x7f0000000180)='./bus\x00', 0x0)
+mkdirat(r0, &(0x7f0000000280)='./bus/file0\x00', 0x0)
+renameat2(r0, &(0x7f00000004c0)='./file0\x00', r0, &(0x7f0000000500)='./bus/file0/file0\x00', 0x0)`,
+			subsystems: []string{"ntfs3", "vfs"},
+		},
+	}
+
+	for i, test := range tests {
+		ret := extractor.Extract(&subsystem.Crash{
+			OS:          targets.Linux,
+			GuiltyFiles: []string{test.guilty},
+			SyzRepro:    test.prog,
+		})
+		sort.Strings(ret)
+		if !reflect.DeepEqual(ret, test.subsystems) {
+			t.Fatalf("test #%d: got %+v, wanted %+v", i, ret, test.subsystems)
+		}
+	}
+}
+
+func TestFsSubsystemFlow(t *testing.T) {
+	// Test that we can do the following:
+	// 1. Delay the reporting of possible vfs bugs until we have found a reproducer.
+	// 2. Once the reproducer comes, extract the extra subsystems and report it.
+
+	c := NewCtx(t)
+	defer c.Close()
+
+	client := c.makeClient(clientPublicFs, keyPublicFs, true)
+	build := testBuild(1)
+	client.UploadBuild(build)
+
+	// A. Make sure non-fs bugs are not affected.
+	// -----------------------------------------
+
+	crash := testCrash(build, 1)
+	crash.Title = "WARNING: abcd"
+	crash.Log = []byte("log log log")
+	crash.GuiltyFiles = []string{"kernel/kernel.c"}
+	crash.Maintainers = []string{"maintainer@kernel.org"}
+	client.ReportCrash(crash)
+
+	// We skip the first stage and report the bug right away.
+	reply := c.pollEmailBug()
+	c.expectEQ(reply.Subject, "[syzbot] WARNING: abcd")
+
+	// B. Send a non-vfs bug without a reproducer.
+	// -----------------------------------------
+
+	crash = testCrash(build, 2)
+	crash.Title = "WARNING in nilfs_dat_commit_end"
+	crash.GuiltyFiles = []string{"fs/nilfs2/dat.c"}
+	crash.Log = []byte("log log log")
+	crash.Maintainers = []string{"maintainer@kernel.org"}
+	client.ReportCrash(crash)
+
+	reply = c.pollEmailBug()
+	// We have no means to determine the subsystem now, so it shouldn't be anywhere.
+	c.expectEQ(reply.Subject, "[syzbot] WARNING in nilfs_dat_commit_end")
+
+	// C. Send a possibly vfs bug without a reproducer.
+	// -----------------------------------------
+
+	crash = testCrash(build, 3)
+	crash.Title = "WARNING in do_mkdirat"
+	crash.GuiltyFiles = []string{"fs/namei.c"}
+	crash.Log = []byte("log log log")
+	crash.Maintainers = []string{"maintainer@kernel.org"}
+	client.ReportCrash(crash)
+
+	// As there's no other information, the bug is left at the first reporting.
+	c.client.pollNotifs(0)
+	vfsBug := client.pollBug()
+
+	// D. Now report a reproducer for the (C) bug that does image mounting.
+	// -----------------------------------------
+
+	crash = testCrash(build, 4)
+	crash.Title = "WARNING in do_mkdirat"
+	crash.GuiltyFiles = []string{"fs/namei.c"}
+	crash.Log = []byte("log log log")
+	// nolint: lll
+	crash.ReproSyz = []byte(`syz_mount_image$ntfs3(&(0x7f0000000240), &(0x7f000001f3c0)='./file0\x00', 0xc40, &(0x7f00000005c0)=ANY=[@ANYBLOB="0032"], 0x3, 0x1f398, &(0x7f000003e7c0)="111")
```

Comment:
I was expecting this test to modify the C crash to add the reproducer - is that not how this system works?

Reply:
If a reproducer is found, it's always reported as a new crash. At least that's how syzbot works now.